go/pkg/services/log_collector/server.go

```diff
@@ -16,36 +16,507 @@ package log_collector
+		}, nil
+	}
+
+	// unmarshal
+	if err := json.Unmarshal(stateFileBytes, state); err != nil {
+		return nil, errors.Wrap(err, "Failed to unmarshal state file")
+	}
+
+	return state, nil
+}
+
+// writeStateToFile writes the state to file
+func (lcs *LogCollectorServer) writeStateToFile(ctx context.Context, state *State) error {
+
+	// marshal state file
+	encodedState, err := json.Marshal(state)
+	if err != nil {
+		return errors.Wrap(err, "Failed to encode state file")
+	}
+
+	// get lock, unlock later
+	lcs.stateFileLock.Lock()
+	defer lcs.stateFileLock.Unlock()
+
+	// write to file
+	stateFilePath := lcs.resolveStateFilePath()
+	return common.WriteToFile(ctx, lcs.Logger, stateFilePath, encodedState, false)
+}
+
+// resolvePodLogFilePath returns the path to the pod log file
+func (lcs *LogCollectorServer) resolvePodLogFilePath(runId, podName string) string {
+	return path.Join(lcs.baseDir, "logs", fmt.Sprintf("%s_%s.log", runId, podName))
+}
+
+func (lcs *LogCollectorServer) resolveStateFilePath() string {
+	return path.Join(lcs.baseDir, "state.json")
+}
+
+func (lcs *LogCollectorServer) getLogFilePath(runId string) (string, error) {
+
+	logFilePath := ""
+	var latestModTime time.Time
+
+	// list all files in base directory
+	err := filepath.Walk(lcs.baseDir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		// if file name starts with run id, it's a log file
+		if strings.HasPrefix(info.Name(), runId) {
+
+			// if it's the first file, set it as the log file path
+			// otherwise, check if it's the latest modified file
+			if logFilePath == "" || info.ModTime().After(latestModTime) {
+				logFilePath = path
+				latestModTime = info.ModTime()
+			}
+		}
+
+		return nil
+	})
+	if err != nil {
+		return "", errors.Wrap(err, "Failed to list files in base directory")
+	}
+
+	// if no log file was found, return error
+	if logFilePath == "" {
+		return "", errors.Errorf("Failed to find log file for run id %s", runId)
+	}
+
+	return logFilePath, nil
+}
+
+func (lcs *LogCollectorServer) validateOffsetAndSize(offset uint64, size uint64, fileSize uint64) (uint64, uint64) {
+
+	// set size to fileSize if size is bigger than fileSize
+	if size > fileSize {
+		size = fileSize
+	}
+
+	// set size to file size - offset if size is bigger than file size - offset
+	if size > fileSize-offset {
+		size = fileSize - offset
+	}
+
+	// if size is zero, read the whole file
+	if size == 0 {
+		size = fileSize
+	}
+
+	// if offset is bigger than file size, set offset to 0
+	if offset > fileSize {
+		offset = 0
+	}
+
+	return offset, size
+}
+
+// updateStateFile periodically checks if changes were made to the state, and updates them in the state file
+func (lcs *LogCollectorServer) updateStateFile(ctx context.Context, state *State) {
+
+	for {
+
+		// get state file
+		currentState := lcs.getState()
+
+		// if state changed, write it to file
+		if !reflect.DeepEqual(currentState, state) {
```

Comment:
these will always be equal. 
lcs.getState() is the current state, and `state *State` is the pointer to the state in the LogCollectorServer which is also the current state.

You need this method to hold 2 variables of current and previous and update them with each iteration (no need to get the state pointer in the parameters)